src/java.base/share/classes/java/util/concurrent/Executors.java

```diff
@@ -824,14 +823,28 @@ public <T> T invokeAny(Collection<? extends Callable<T>> tasks,
         }
     }
 
-    private static class FinalizableDelegatedExecutorService
+    /**
+     * A DelegatedExecutorService that uses a Cleaner to shut down the underlying
+     * ExecutorService when the wrapper becomes phantom reachable.
+     */
+    private static class AutoShutdownDelegatedExecutorService
             extends DelegatedExecutorService {
-        FinalizableDelegatedExecutorService(ExecutorService executor) {
+        private final Cleanable cleaner;
```

Comment:
"cleaner" is a Cleanable, not a Cleaner. We should consistently use (as we do elsewhere in openjdk)

```
 private final Cleanable cleanable;
```

and fix the others to be consistent:

```
./sun/nio/ch/DatagramChannelImpl.java:115:    private final Cleanable cleaner;
./sun/nio/ch/NioSocketImpl.java:106:    private Cleanable cleaner;
./java/util/concurrent/Executors.java:832:        private final Cleanable cleaner;
```

(BTW: I find Cleaners much harder to use than old finalize, and it looks like I'm not the only one!)

Reply:
I'm rename the Cleaner in AutoShutdownDelegatedExecutorService but I'd prefer not rename fields in other areas, at least not part of this change.

> (BTW: I find Cleaners much harder to use than old finalize, and it looks like I'm not the only one!)

Finalization has many flaws and I think we could add that it was way too easy to add a finalize method without thinking too much about the implications. So yes, Cleaners do force you to separate out the resource and think about concurrency. It is an advanced feature, to be used very sparely, and we should continue to developers to close the resource explicitly.